Store migration support for block devices

ragadeeshu opened the pull request:
Make store migration work for custom IO integrations, by accessing store files through the page cache. Including for file operations such as deletes, copying and moving files.

This work includes a new `StoreMigrationIT` test, which searches for all available store formats, and tests migration for all supported from-to format tuples. This test is used in the block device integration code to verify that migration works there as well.

The page cache based file operations only take effect for formats from 2.3 and onward.

ragadeeshu committed:
Initial support for blockdevice store migration

Supports migrating nodes and relationships. This is achieved by making the StoreMigrator partly use the page cache for migration. Also adds support for a BatchingneoStores that uses an external page cache.

ragadeeshu committed:
Better support for block device store migration

This is achieved by also using the page cache for relevant parts of the migration process, in order to reach files only reachable by the page cache.

ragadeeshu committed:
Make copyTo work with adversarial page cursors

ragadeeshu committed:
Use TestDirectory for the StoreMigrationIT

In order to make this work, the TestDirectory was opened up a bit, so it can
be used to prepare a test directory, without being properly initialised as a
TestRule. The reason for this is that we needed to use it from the parameter
generating method of a @Parameterised test, and that is called before even
the initialisation of @ClassRules.

ragadeeshu committed:
Cleanup and test fixes for block device migration

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java:
```diff
@@ -466,6 +477,29 @@ private void migrateWithBatchImporter( File storeDir, File migrationDir, long la
+                            storeFile -> storeFile.fileName( StoreFileType.STORE ).equals( fh.getFile().getName() );
+                    if ( storesToDeleteFromMigratedDirectory.stream().anyMatch( predicate ) )
+                    {
+                        final Optional<PagedFile> optionalPagedFile = pageCache.getExistingMapping( fh.getFile() );
```
I don't see how there can still be a mapping on any file here. All of the paged files should be closed once the import finishes.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreUpgrader.java:
```diff
@@ -256,6 +256,20 @@ private void migrateToIsolatedDirectory( File storeDir, File migrationDirectory,
     {
+        // We use the page cache here to make sure that the migration directory is clean even if we are using a block
+        // device.
+        try
```
Here we delete files through `pageCache` first and `fileSystem` second.
In `StoreMigrator` it is the other way around. 
I think it make sense to use the same pattern between the two. Especially since most difficulties around adapting Neo to behave with blockdevice is around file operations.

Maybe consider using `FileHandle.HANDLE_DELETE`?

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreUpgrader.java:
```diff
@@ -256,6 +256,20 @@ private void migrateToIsolatedDirectory( File storeDir, File migrationDirectory,
+        catch ( IOException e )
+        {
+            // This means that we had no files to clean, this is fine.
+        }
```
We get this `IOException` from `SingleFilePageSwapperFactory.streamFilesRecursiveInner(...)` if `migrationDirectory` does not exist. I would prefer if we specifically caught the `NoSuchFileException` that it throws instead to not hide other possible `IOExceptions`, for example if some file is already mapped we get `FileIsMappedException`.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java:
```diff
@@ -466,6 +477,29 @@ private void migrateWithBatchImporter( File storeDir, File migrationDir, long la
+            }
+            catch ( IOException e )
+            {
+                // This means that we had no files only present in the page cache, this is fine.
```
Same comment as for `StoreUpgrader`, even though we explicitly `close` mapped files, there could be other `IOException` that we do not want to hide.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java:
```diff
@@ -520,6 +582,23 @@ private void prepareBatchImportMigration( File storeDir, File migrationDir, Reco
+            neoStores.getPropertyStore();
+            neoStores.getRelationshipGroupStore();
+            neoStores.getRelationshipStore();
+            neoStores.getSchemaStore();
```
Are all of these calls here to make us fail fast if some store is present? In that case 👍

ragadeeshu replied:
These calls are there in order to explicitly create the stores, I think. This method has just been migrated from `BatchingNeoStores` since only this is the only class that calls it.

burqen replied:
Alright, that was my first thought as well but i couldn't see any creation being done in those code paths.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java:
```diff
@@ -634,6 +714,30 @@ public void moveMigratedFiles( File migrationDir, File storeDir, String versionT
+        }
+        catch ( IOException e )
+        {
+            //This means that we had no files only present in the page cache, this is fine.
```
Same `IOException` comment again.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java:
```diff
@@ -634,6 +714,30 @@ public void moveMigratedFiles( File migrationDir, File storeDir, String versionT
                 ExistingTargetStrategy.OVERWRITE, // allow to overwrite target files
                 StoreFileType.values() );
+        // Since some of the files might only be accessible through the page cache (i.e. block devices), we also try to
+        // move the files with the page cache.
```
I like this pattern of first trying to finish the job with `FileSystemAbstraction` and then "clean up" what's left with `PageCache`.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java:
```diff
@@ -634,6 +714,30 @@ public void moveMigratedFiles( File migrationDir, File storeDir, String versionT
+        try
+        {
+            Iterable<FileHandle> fileHandles = pageCache.streamFilesRecursive( migrationDir )::iterator;
+            for ( FileHandle fh : fileHandles )
```
This block has a lot in common with same block in `migrateWithBatchImporter` when deleting files from `migrationDir`. Maybe the behaviour could be extracted into a helper method by extracting the action (delete, rename) and a `Supplier<Stream<StoreFile>>` to test the predicate for.

ragadeeshu replied:
This, along with consolidating all/most of our file helper classes, should probably be done separately.

burqen replied:
I was thinking of extracting it only for this class, but a larger refactoring might be a good idea. If you plan to do this, then nevermind, otherwise feel free to address it for this class in this PR.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java:
```diff
@@ -357,6 +367,7 @@ private void removeDuplicateEntityProperties( File storeDir, File migrationDir,
             SchemaIndexProvider schemaIndexProvider, RecordFormats oldFormat )
             throws IOException
     {
+        CustomIOConfigValidator.assertCustomIOConfigNotUsed( config, CUSTOM_IO_EXCEPTION_MESSAGE );
```
Can we please move this assertion to before calling this method and add a comment about it, like we do in `rebuildCounts` so it becomes:
```
if ( versionToMigrateFrom.equals( StandardV2_1.STORE_VERSION ) )
{
    // These versions are not supported for block devices.
    CustomIOConfigValidator.assertCustomIOConfigNotUsed( config, CUSTOM_IO_EXCEPTION_MESSAGE );
    removeDuplicateEntityProperties( storeDir, migrationDir, pageCache, schemaIndexProvider, oldFormat );
}
```

Also, why do we not check for `StandardV2_2.STORE_VERSION` like we do in `rebuildCounts`?

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreUpgrader.java:
```diff
@@ -116,8 +118,6 @@ public void migrateIfNeeded( File storeDirectory)
         }
 
-        CustomIOConfigValidator.assertCustomIOConfigNotUsed( config, CUSTOM_IO_EXCEPTION_MESSAGE );
-
```
Instead of removing this from here and check it later in `StoreMigrator`, does it make sense to instead fail as fast as possible here as soon and `versionToMigrateFrom` is known?
I mean inside the `if ( MigrationStatus.X.isNeededFor( migration status ) )` blocks.

ragadeeshu replied:
One argument against this is that the `StoreUpgrader` would need to become aware of store versions, instead of that being isolated to `StoreMigrator`

burqen replied:
It's actually `StoreUpgrader` that reads store version and passes it to `StoreMigrator`.

ragadeeshu replied:
@burqen I meant checking against specific versions, importing `StandardV` classes, etc. `StoreUpgrader` is currently free of such logic, while `StoreMigrator` is full of it already. `StoreMigrater` is also the only `Migrator` implementation that is incompatible with custom IO configurations.

burqen replied:
Ah, then would agree that it makes more sense keeping it as it is.

burqen commented on community/neo4j/src/test/java/upgrade/PlatformConstraintStoreUpgradeTest.java:
```diff
@@ -68,7 +68,7 @@ public void shouldFailToStartWithCustomIOConfigurationTest() throws IOException
         catch ( RuntimeException ex )
         {
-            assertEquals( StoreUpgrader.CUSTOM_IO_EXCEPTION_MESSAGE, ex.getCause().getCause().getMessage() );
+            assertEquals( StoreMigrator.CUSTOM_IO_EXCEPTION_MESSAGE, ex.getCause().getCause().getMessage() );
```
Should we not also check for `StandardV2_1.STORE_VERSION` here?

ragadeeshu replied:
Fixed, also added 2.0

burqen reviewed (changes requested):
First review round finished. Have some concerns about Exception handling and some small comments to improve readability.

burqen commented:
@ragadeeshu Comments ready for you 👍

ragadeeshu committed:
Addressing pull request comments.

ragadeeshu commented:
@burqen Please have a look again :)

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreUpgrader.java:
```diff
@@ -256,31 +258,27 @@ private void migrateToIsolatedDirectory( File storeDir, File migrationDirectory,
         {
-            Iterable<FileHandle> fileHandles = pageCache.streamFilesRecursive( migrationDirectory )::iterator;
-            for ( FileHandle fh : fileHandles )
+            if ( migrationDirectory.exists() )
```
I missed this during my first review apparently, sorry.
We should work towards using `FileSystemAbstraction` whenever possible, rather than mixing between `FSA`, `FileUtils` and `File#...`. Feel free to use `fileSystem.fileExists( migrationDirectory )` here.

burqen reviewed (changes requested):
Second review finished. Changes look good. Approved once I get an answer on the remaining questions.

burqen commented:
@ragadeeshu Second review finished. There are some test failures that seem related to these changes as well.

ragadeeshu committed:
Make StoreUpgraderTest more specific

failMigrationWhenFailDuringTransactionInformationRetrieval now only throws for in the relevant place.
